### ISSUE TITLE
Edge case bug in VXCao_empty_blocks_cider

### ISSUE THREAD

**Issue by kylebystrom** [pyscf/pyscf#1191]
I think that there is an edge case bug in the empty block screening routine for numerical integration of the XC kernel, but I am having trouble finding an edge case where it causes numerically significant issues.

Starting on line 43 of `nr_numint.c` (https://github.com/pyscf/pyscf/blob/50c6918f77d10692fcd1aa8e95f9eb89fb4236aa/pyscf/lib/dft/nr_numint.c#L43), the following changes should be made:
```c
for (bas_id = sh0; bas_id < sh1; bas_id++) {
        empty[box_id] &= !non0table[bas_id];
        if (ao_loc[bas_id] == bound) {
                has0 |= empty[box_id];
                box_id++;
                bound += BOXSIZE;
                empty[box_id] = 1;
        } else if (ao_loc[bas_id] > bound) {
                has0 |= empty[box_id];
                box_id++;
                bound += BOXSIZE;
                empty[box_id] = !non0table[bas_id];
        }
}
```
should be
```c
for (bas_id = sh0; bas_id < sh1; bas_id++) {
        if (ao_loc[bas_id] == bound) {
                has0 |= empty[box_id];
                box_id++;
                bound += BOXSIZE;
                empty[box_id] = 1;
        } 
        empty[box_id] &= !non0table[bas_id];
        if (ao_loc[bas_id+1] > bound) {
                has0 |= empty[box_id];
                box_id++;
                bound += BOXSIZE;
                empty[box_id] = !non0table[bas_id];
        }
}
```

The reason for these changes is that if `ao_loc[bas_id]` is equal to `bound`, its orbitals contribute to `empty[box_id+1]`, not `empty[box_id]`, so the `box_id` should be incremented before handling `bas_id`. If `ao_loc[bas_id+1]` is greater than `bound`, then the orbitals at `ao_loc[bas_id]` contribute to both `empty[box_id]` and `empty[box_id+1]`, so it is necessary to check `ao_loc[bas_id+1] > bound` rather than `ao_loc[bas_id] > bound` in the second if statement.

Unfortunately, I found this bug while writing some custom code that uses PySCF's `VXCao_empty_blocks` function, and I have not found an example using just PySCF that runs into the edge cases where this bug causes significant problems. Does anyone have suggestions for reproducing this issue using the main code? If so, I can draw up some tests and make a PR for the changes.

**Comment by sunqm** [comment]
For small systems,  dgemm between the matrices are called. The bug never be noticed in these small tests.

The check for boundary of `ao_loc` and `bound` should be carried out between next basis and bound. IMO, the fix as below is more clear

```
for (bas_id = sh0; bas_id < sh1; bas_id++) {
        empty[box_id] &= !non0table[bas_id];
        if (ao_loc[bas_id+1] == bound) {
                has0 |= empty[box_id];
                box_id++;
                bound += BOXSIZE;
                empty[box_id] = 1;
        } else if (ao_loc[bas_id+1] > bound) {
                has0 |= empty[box_id];
                box_id++;
                bound += BOXSIZE;
                empty[box_id] = !non0table[bas_id];
        }
}
```

**Comment by kylebystrom** [comment]
I don't think that version will work. If `nao % BOXSIZE == 0`, then for the last iteration of the for loop,
```
bas_id = sh1-1
box_id = nbox - 1
bound = nao = ao_loc[bas_id+1] = ao_loc[sh1]
```
Therefore, the first conditional will be true, resulting in `box_id` being incremented to `nbox` and the invalid write `empty[nbox] = 1` being performed. If the check is performed for `ao_loc[bas_id]` before calling `empty[box_id] &= !non0table[bas_id];`, then this issue will not occur.

**Comment by kylebystrom** [comment]
Fixed by #1274
